modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -266,34 +267,41 @@ object ConsensusStateUpdater {
                     _ <- consensusStorage.addEvents(returnedPeerEvents)
                     hash <- artifact.hashF
                     effect = gossip.spread(ConsensusPeerDeclaration(key, Proposal(hash)))
-                  } yield (ProposalInfo[Artifact](artifact, hash).some, effect)
-                }.map(_.getOrElse((none[ProposalInfo[Artifact]], Applicative[F].unit))).map {
-                  case (maybeProposalInfo, effect) =>
-                    val newState = state.copy(status = CollectingProposals[Artifact](majorityTrigger, maybeProposalInfo))
+                    _ <- gossip.spreadCommon(ConsensusArtifact(key, artifact))
+                  } yield (ProposalInfo[Artifact](artifact, hash).some, facilityInfo.lastSignedArtifact.some, effect)
+                }.map(_.getOrElse((none[ProposalInfo[Artifact]], none[Signed[Artifact]], Applicative[F].unit))).map {
+                  case (maybeProposalInfo, lastSignedArtifact, effect) =>
+                    val newState =
+                      state.copy(status = CollectingProposals[Artifact](majorityTrigger, maybeProposalInfo, lastSignedArtifact))
                     (newState, effect)
                 }
           }
-        case CollectingProposals(majorityTrigger, maybeProposalInfo) =>
+
+        case CollectingProposals(majorityTrigger, maybeProposalInfo, maybeLastArtifact) =>
           val maybeAllProposals = state.facilitators
             .traverse(resources.peerDeclarationsMap.get)
             .flatMap(_.traverse(_.proposal.map(_.hash)))
 
           maybeAllProposals.flatTraverse { allProposals =>
-            pickMajority(allProposals).traverse { majorityHash =>
-              val newState = state.copy(status = CollectingSignatures[Artifact](majorityHash, majorityTrigger))
-              val effect = maybeProposalInfo.traverse { proposalInfo =>
-                Signature.fromHash(keyPair.getPrivate, majorityHash).flatMap { signature =>
-                  gossip.spread(ConsensusPeerDeclaration(key, MajoritySignature(signature)))
-                } >> Applicative[F].whenA(majorityHash === proposalInfo.proposalArtifactHash) {
-                  gossip.spreadCommon(ConsensusArtifact(key, proposalInfo.proposalArtifact))
-                } >> Metrics[F].recordDistribution(
-                  "dag_consensus_proposal_affinity",
-                  proposalAffinity(allProposals, proposalInfo.proposalArtifactHash)
-                )
-              }.void
-              (newState, effect).pure[F]
+            (maybeLastArtifact match {
+              case Some(lastSignedArtifact) => pickValidatedMajority(lastSignedArtifact, majorityTrigger, resources)(allProposals)
```

Comment:
Let's use either `.map` combined with `.getOrElse` instead of the lengthy pattern match.

```
maybeLastArtifact.map { lastSignedArtifact =>
 ...
}.getOrElse(...)
```